controllers/machine_controller_noderef_test.go

```diff
@@ -20,105 +20,125 @@ import (
-				ProviderID: "gce://us-central1/id-node-2",
-			},
+	ns, err := testEnv.CreateNamespace(ctx, "test-get-node")
+	g.Expect(err).ToNot(HaveOccurred())
+
+	// Set up cluster to test against.
+	testCluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-get-node-",
+			Namespace:    ns.Name,
 		},
 	}
 
-	client := fake.NewClientBuilder().WithObjects(nodeList...).Build()
+	g.Expect(testEnv.Create(ctx, testCluster)).To(BeNil())
+	g.Expect(testEnv.CreateKubeconfigSecret(ctx, testCluster)).To(Succeed())
+	defer func(do ...client.Object) {
+		g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+	}(ns, testCluster)
 
 	testCases := []struct {
-		name       string
-		providerID string
-		expected   *corev1.ObjectReference
-		err        error
+		name            string
+		node            *corev1.Node
+		providerIDInput string
 	}{
 		{
-			name:       "valid provider id, valid aws node",
-			providerID: "aws:///id-node-1",
-			expected:   &corev1.ObjectReference{Name: "node-1"},
-		},
-		{
-			name:       "valid provider id, valid aws node",
-			providerID: "aws:///id-node-2",
-			expected:   &corev1.ObjectReference{Name: "node-2"},
+			name: "full providerID matches",
+			node: &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node-1",
+				},
+				Spec: corev1.NodeSpec{
+					ProviderID: "aws://us-east-1/id-node-1",
+				},
+			},
+			providerIDInput: "aws://us-east-1/id-node-1",
 		},
 		{
-			name:       "valid provider id, valid gce node",
-			providerID: "gce:///id-node-2",
-			expected:   &corev1.ObjectReference{Name: "gce-node-2"},
+			name: "aws prefix: cloudProvider and ID matches",
+			node: &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node-2",
+				},
+				Spec: corev1.NodeSpec{
+					ProviderID: "aws://us-west-2/id-node-2",
+				},
+			},
+			providerIDInput: "aws:///id-node-2",
 		},
 		{
-			name:       "valid provider id, no node found",
-			providerID: "aws:///id-node-100",
-			expected:   nil,
-			err:        ErrNodeNotFound,
+			name: "gce prefix, cloudProvider and ID matches",
+			node: &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "gce-node-2",
+				},
+				Spec: corev1.NodeSpec{
+					ProviderID: "gce://us-central1/id-node-2",
+				},
+			},
+			providerIDInput: "gce:///id-node-2",
 		},
 	}
 
-	for _, test := range testCases {
-		t.Run(test.name, func(t *testing.T) {
-			gt := NewWithT(t)
-			providerID, err := noderefutil.NewProviderID(test.providerID)
-			gt.Expect(err).NotTo(HaveOccurred(), "Expected no error parsing provider id %q, got %v", test.providerID, err)
-
-			node, err := r.getNode(ctx, client, providerID)
-			if test.err == nil {
-				g.Expect(err).To(BeNil())
-			} else {
-				gt.Expect(err).NotTo(BeNil())
-				gt.Expect(err).To(Equal(test.err), "Expected error %v, got %v", test.err, err)
-			}
+	nodesToCleanup := make([]client.Object, 0, len(testCases))
+	for _, tc := range testCases {
+		g.Expect(testEnv.Create(ctx, tc.node)).To(BeNil())
```

Comment:
Just out of curiosity (and I'm new to most of this, so I could be wrong about anything I assume :)). 

We're using the same testEnv for multiple tests (in parallel?). Usually the tests are separated from each other because we deploy resources into a test-specific namespace. Could it lead to strange side effects in other tests when we deploy cluster-wide resources like Nodes?

Reply:
Sounds reasonable to me, I renamed the nodes to be test-name-prefixed as a good practice. Thanks!